test/tla/tutorials/randomized/test_erc20.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
+#
+# Modeling ERC20 tokens of Ethereum and the Approve-TransferFrom Attack:
+#
+# EIP-20: https://eips.ethereum.org/EIPS/eip-20
+#
+# Attack scenario:
+#   https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit#
+#
+# Our testing framework is designed towards checking the protocol features.
+# We do not model 256-bit integers here, as we are not interested in overflows.
+# 
+# Igor Konnov, Informal Systems, 2021-2022
+
+
+import unittest
+
+from hypothesis import given, strategies as gen
+from hypothesis.stateful import Bundle, RuleBasedStateMachine
+from hypothesis.stateful import rule, consumes, invariant, initialize
+from hypothesis import assume, settings, event, Verbosity
+
+# The list of addresses to use. We could use real addresses here,
+# but simple readable names are much nicer.
+ADDR = [ "Alice", "Bob", "Eve" ]
+
+# We restrict the amounts to a small range, to avoid too much randomness
+AMOUNTS = range(0, 20)
+
+
+class TransferTx:
+    """An instance of transfer"""
+
+    def __init__(self, sender, toAddr, value):
+        self.tag = "transfer"
+        self.sender = sender
+        self.toAddr = toAddr
+        self.value = value
+
+class TransferFromTx:
+    """An instance of transferFrom"""
+
+    def __init__(self, sender, fromAddr, toAddr, value):
+        self.tag = "transferFrom"
+        self.sender = sender
+        self.fromAddr = fromAddr
+        self.toAddr = toAddr
+        self.value = value
+
+class ApproveTx:
+    """An instance of approve"""
+
+    def __init__(self, sender, spender, value):
+        self.tag = "approve"
+        self.sender = sender
+        self.spender = spender
+        self.value = value
+
+
+class Erc20Simulator(RuleBasedStateMachine):
+    """
+    Model the behavior of the ERC20 API in terms of stateful testing.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    # This bundle contains the generated transactions
+    # that are to be processed
+    pendingTxs = Bundle("pendingTxs")
+
+    @initialize(amounts=gen.lists(gen.sampled_from(AMOUNTS),
+                min_size=len(ADDR),
+                max_size=len(ADDR)))
+    def init(self, amounts):
+        # balance of every account
+        self.balanceOf = {
+            addr: amount for (addr, amount) in zip(ADDR, amounts)
+        }
+        # approvals from senders to spenders
+        self.allowance = {
+            (sender, spender): 0 for sender in ADDR for spender in ADDR
+        }
+        # history variables that we need to express the invariants
+        self.pendingTxsShadow = set()
+        self.lastTx = None
```

Comment:
What does `lastTx` actually capture? It seems to be more complex than "the last committed transaction", because we reset it to `None` in the `submit_*` methods.

I was puzzled by this quite a bit, and it's not explained in the prose either.
Please add a more meaningful comment.